wgengine/monitor/monitor_windows.go

```diff
@@ -7,273 +7,149 @@ package monitor
+	defer func() {
+		m.mu.Lock()
+		m.messagec = nil
+		m.mu.Unlock()
+	}()
 
-		// Something changed, so assume Windows is about to
-		// discover its new proxy settings from WPAD, which
-		// seems to take a bit. Poll heavily for awhile.
-		m.inFastPoll = true
-		m.pollTicker.Reset(pollIntervalFast)
-	}
 	m.mu.Unlock()
 
-	return unspecifiedMessage{}, nil
-}
+	t0 := time.Now()
 
-func notifyAddrChange() (h windows.Handle, o *windows.Overlapped, cancel func(), err error) {
-	return callNotifyProc(notifyAddrChangeProc)
+	select {
+	case msg := <-m.messagec:
+		m.logf("got windows change event after %v: evt=%s", time.Since(t0).Round(time.Millisecond), msg.eventType)
+		return msg, nil
+	case <-m.ctx.Done():
+		return nil, errClosed
+	}
 }
 
-func notifyRouteChange() (h windows.Handle, o *windows.Overlapped, cancel func(), err error) {
-	return callNotifyProc(notifyRouteChangeProc)
+// unicastAddressChanged is the callback we register with Windows to call when unicast address changes.
+func (m *winMon) unicastAddressChanged(_ winipcfg.MibNotificationType, _ *winipcfg.MibUnicastIPAddressRow) {
+	// start a goroutine to finish our work, to return to Windows out of this callback
+	go m.somethingChanged("addr")
 }
 
-func callNotifyProc(p *syscall.LazyProc) (h windows.Handle, o *windows.Overlapped, cancel func(), err error) {
-	o = new(windows.Overlapped)
-
-	// TODO(bradfitz): understand why this if-false code doesn't
-	// work, even though the docs online suggest we should pass an
-	// event in the overlapped.Hevent field.
-	// The docs at
-	// https://docs.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-overlapped
-	// says that o.HEvent can be zero, though, which seems to work.
-	// Note that the returned windows.Handle returns the same value for both
-	// notifyAddrChange and notifyRouteChange, which is why our caller needs
-	// to look at the returned Overlapped's Internal field to see which case
-	// fired. That's also worth understanding more.
-	// See crawshaw's comment at https://github.com/tailscale/tailscale/pull/944#discussion_r526469186
-	// too.
-	if false {
-		evt, err := windows.CreateEvent(nil, 0, 0, nil)
-		if err != nil {
-			return 0, nil, nil, err
-		}
-		o.HEvent = evt
-	}
+// routeChanged is the callback we register with Windows to call when route changes.
+func (m *winMon) routeChanged(_ winipcfg.MibNotificationType, _ *winipcfg.MibIPforwardRow2) {
+	// start a goroutine to finish our work, to return to Windows out of this callback
+	go m.somethingChanged("route")
+}
 
-	r1, _, e1 := syscall.Syscall(p.Addr(), 2, uintptr(unsafe.Pointer(&h)), uintptr(unsafe.Pointer(o)), 0)
+// somethingChanged gets called from OS callbacks whenever address or route changes.
+func (m *winMon) somethingChanged(evt string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
 
-	// We expect ERROR_IO_PENDING.
-	if syscall.Errno(r1) != windows.ERROR_IO_PENDING {
-		return 0, nil, nil, e1
+	if m.messagec == nil {
```

Comment:
Delete these 4 lines so we don't miss a message between Receive calls. If it's closed, line 150 will catch it.

Also, no need for the m.mu.Lock/Unlock above then, if you restore messagec to be created once at the beginning like it was previously.